# **Boosting Wav2Vec2 with n-grams in 🤗 Transformers**

**Wav2Vec2** is a popular pre-trained model for speech recognition. Released in [September 2020](https://ai.facebook.com/blog/wav2vec-20-learning-the-structure-of-speech-from-raw-audio/) by Meta AI Research, the novel architecture catalyzed progress in self-supervised pretraining for speech recognition, *e.g.* [*G. Ng et al.*, 2021](https://arxiv.org/pdf/2104.03416.pdf), [*Chen et al*, 2021](https://arxiv.org/abs/2110.13900), [*Hsu et al.*, 2021](https://arxiv.org/abs/2106.07447) and [*Babu et al.*, 2021](https://arxiv.org/abs/2111.09296). On the Hugging Face Hub, Wav2Vec2's most popular pre-trained checkpoint currently amounts to over [**250,000** monthly downloads](https://huggingface.co/facebook/wav2vec2-base-960h).

Using Connectionist Temporal Classification (CTC), pre-trained Wav2Vec2-like checkpoints are extremely easy to fine-tune on downstream speech recognition tasks.
In a nutshell, fine-tuning pre-trained Wav2Vec2 checkpoints works as follows: 

A single randomly initialized linear layer is stacked on top of the pre-trained checkpoint and trained to classify raw audio input to a sequence of letters. It does so by:

1.  extracting audio representations from the raw audio (using CNN layers),
2. processing the sequence of audio representations with a stack of transformer layers, and,
3. classifying the processed audio representations into a sequence of output letters.

Previously audio classification models required an additional language model (LM) and a dictionary to transform the sequence of classified audio frames to a coherent transcription.
Wav2Vec2's architecture is based on transformer layers, thus giving each processed audio representation context 
from all other audio representations. In addition, 
Wav2Vec2 leverages the [CTC algorithm](https://distill.pub/2017/ctc/) for fine-tuning, which solves the problem of alignment between a varying "input audio length"-to-"output text length" ratio.

Having contextualized audio classifications and no alignment problems, Wav2Vec2 does not require 
an external language model or dictionary to yield acceptable audio transcriptions.

As can be seen in Appendix C of the [official paper](https://arxiv.org/abs/2006.11477), Wav2Vec2 gives impressive downstream performances on [LibriSpeech]((https://huggingface.co/datasets/librispeech_asr)) without using a language model at all. However, from the appendix, it also becomes clear that using Wav2Vec2 in combination with a language model can yield a significant improvement, especially when the model was trained on only 10 minutes of transcribed audio.

Until recently, the 🤗 Transformers library did not offer a simple user interface to decode audio files with a fine-tuned Wav2Vec2 **and** a language model. This has thankfully changed. 🤗 Transformers now offers an easy-to-use integration with *Kensho Technologies'* [pyctcdecode library](https://github.com/kensho-technologies/pyctcdecode). This blog post is a step-by-step **technical** guide to explain how one can create an **n-gram** language model and combine it with an existing fine-tuned Wav2Vec2 checkpoint using 🤗 Datasets and 🤗 Transformers.

We start by:

1. How does decoding audio with an LM differ from decoding audio without an LM?
2. How to get suitable data for a language model?
3. How to build an *n-gram* with KenLM?
4. How to combine the *n-gram* with a fine-tuned Wav2Vec2 checkpoint?

For a deep dive into how Wav2Vec2 functions - which is not necessary for this blog post - the reader is advised to consult the following material:

- [wav2vec 2.0: A Framework for Self-Supervised Learning of Speech Representations](https://arxiv.org/abs/2006.11477)
- [Fine-Tune Wav2Vec2 for English ASR with 🤗 Transformers](https://huggingface.co/blog/fine-tune-wav2vec2-english)
- [An Illustrated Tour of Wav2vec 2.0](https://jonathanbgn.com/2021/09/30/illustrated-wav2vec-2.html)

## **1. Decoding audio data with Wav2Vec2 and a language model**

As shown in 🤗 Transformers [exemple docs of Wav2Vec2](https://huggingface.co/docs/transformers/master/en/model_doc/wav2vec2#transformers.Wav2Vec2ForCTC), audio can be transcribed as follows.

First, we install `datasets` and `transformers`.



In [ ]:
!pip install datasets transformers torchaudio

     |████████████████████████████████| 306 kB 4.9 MB/s 
     |████████████████████████████████| 3.4 MB 50.0 MB/s 
     |████████████████████████████████| 243 kB 71.9 MB/s 
     |████████████████████████████████| 133 kB 66.6 MB/s 
     |████████████████████████████████| 67 kB 4.9 MB/s 
     |████████████████████████████████| 1.1 MB 34.7 MB/s 
     |████████████████████████████████| 596 kB 56.1 MB/s 
     |████████████████████████████████| 895 kB 56.2 MB/s 
     |████████████████████████████████| 3.3 MB 53.5 MB/s 
     |████████████████████████████████| 271 kB 68.0 MB/s 
     |████████████████████████████████| 144 kB 78.7 MB/s 
     |████████████████████████████████| 160 kB 48.0 MB/s 
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


Let's load a small excerpt of the [Librispeech dataset](https://huggingface.co/datasets/librispeech_asr) to demonstrate Wav2Vec2's speech transcription capabilities.

In [ ]:
from huggingface_hub import notebook_login

notebook_login()

Login successful
Your token has been saved to /root/.huggingface/token
Authenticated through git-credential store but this isn't the helper defined on your machine.
You might have to re-authenticate when pushing to the Hugging Face Hub. Run the following command in your terminal in case you want to set this credential helper as the default

git config --global credential.helper store


In [ ]:
from datasets import load_dataset

dataset = load_dataset(
    "mozilla-foundation/common_voice_7_0", "as", data_dir="./cv-corpus-7.0-2021-07-21", use_auth_token=True, split="test")
dataset

Downloading:   0%|          | 0.00/9.79k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.92k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/49.7k [00:00<?, ?B/s]

Using custom data configuration as-08527a4ca7b6e8cd


Downloading:   0%|          | 0.00/34.9M [00:00<?, ?B/s]

0 examples [00:00, ? examples/s]

0 examples [00:00, ? examples/s]

0 examples [00:00, ? examples/s]

0 examples [00:00, ? examples/s]

0 examples [00:00, ? examples/s]

Dataset common_voice downloaded and prepared to /root/.cache/huggingface/datasets/mozilla-foundation___common_voice/as-08527a4ca7b6e8cd/7.0.0/33e08856cfa0d0665e837bcad73ffd920a0bc713ce8c5fffb55dbdf1c084d5ba. Subsequent calls will reuse this data.


Dataset({
    features: ['client_id', 'path', 'audio', 'sentence', 'up_votes', 'down_votes', 'age', 'gender', 'accent', 'locale', 'segment'],
    num_rows: 281
})

We can pick one of the 73 audio samples and listen to it.

In [ ]:
import IPython.display as ipd

index = 3
audio_sample = dataset[index]
print(audio_sample["sentence"].lower())
ipd.Audio(data=audio_sample["audio"]["array"], autoplay=True, rate=audio_sample["audio"]["sampling_rate"])

"কানায়ে ৰজাৰ আগত আঁঠু লৈ ক'লে।"


Having chosen a data sample, we now load the fine-tuned model and processor.

In [ ]:
from transformers import Wav2Vec2Processor, Wav2Vec2ForCTC

repo_name = 'infinitejoy/wav2vec2-large-xls-r-300m-assamese'

processor = Wav2Vec2Processor.from_pretrained(repo_name)
model = Wav2Vec2ForCTC.from_pretrained(repo_name)

Downloading:   0%|          | 0.00/214 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/304 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/729 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/23.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/502 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.02k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.18G [00:00<?, ?B/s]

Next, we process the data

In [ ]:
inputs = processor(audio_sample["audio"]["array"], sampling_rate=16_000, return_tensors="pt")

forward it to the model

In [ ]:
import torch

with torch.no_grad():
  logits = model(**inputs).logits

/usr/local/lib/python3.7/dist-packages/transformers/models/wav2vec2/modeling_wav2vec2.py:1093: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  return (input_length - kernel_size) // stride + 1


and decode it

In [ ]:
predicted_ids = torch.argmax(logits, dim=-1)
transcription = processor.batch_decode(predicted_ids)

transcription[0].lower()

'ছক'

In [ ]:
import torchaudio

# processor = Wav2Vec2Processor.from_pretrained("airesearch/wav2vec2-large-xlsr-53-th")
# model = Wav2Vec2ForCTC.from_pretrained("airesearch/wav2vec2-large-xlsr-53-th")

#function to resample to 16_000
def speech_file_to_array_fn(batch, 
                            text_col="sentence", 
                            fname_col="path",
                            resampling_to=16000):
    speech_array, sampling_rate = torchaudio.load(batch[fname_col])
    resampler=torchaudio.transforms.Resample(sampling_rate, resampling_to)
    batch["speech"] = resampler(speech_array)[0].numpy()
    batch["sampling_rate"] = resampling_to
    batch["target_text"] = batch[text_col]
    return batch

#get 2 examples as sample input
test_dataset = dataset.map(speech_file_to_array_fn)
inputs = processor(test_dataset["speech"][:2], sampling_rate=16_000, return_tensors="pt", padding=True)

#infer
with torch.no_grad():
    logits = model(inputs.input_values,).logits

predicted_ids = torch.argmax(logits, dim=-1)

print("Prediction:", processor.batch_decode(predicted_ids))
print("Reference:", test_dataset["sentence"][:2])

  0%|          | 0/281 [00:00<?, ?ex/s]

RuntimeError: ignored

In [ ]:
!ls -ltr ~/.cache/huggingface/datasets/mozilla-foundation___common_voice/as-08527a4ca7b6e8cd/7.0.0/33e08856cfa0d0665e837bcad73ffd920a0bc713ce8c5fffb55dbdf1c084d5ba

total 35520
-rw-r--r-- 1 root root 20119920 Jan 21 09:45 common_voice-train.arrow
-rw-r--r-- 1 root root 11121096 Jan 21 09:45 common_voice-test.arrow
-rw-r--r-- 1 root root  3741504 Jan 21 09:45 common_voice-validation.arrow
-rw-r--r-- 1 root root     1496 Jan 21 09:45 common_voice-other.arrow
-rw-r--r-- 1 root root  1365840 Jan 21 09:45 common_voice-invalidated.arrow
-rw-r--r-- 1 root root       50 Jan 21 09:45 LICENSE
-rw-r--r-- 1 root root     3821 Jan 21 09:45 dataset_info.json


Comparing the transcription to the target transcription above, we can see that some words *sound* correct, but are not *spelled* correctly, *e.g.*:

- *christmaus* vs. *christmas*
- *rose* vs. *roast*
- *simalyis* vs. *similes*

Let's see whether combining Wav2Vec2 with an ***n-gram*** lnguage model can help here.

First, we need to install `pyctcdecode` and `kenlm`.

In [ ]:
!pip install https://github.com/kpu/kenlm/archive/master.zip pyctcdecode

     / 541 kB 2.3 MB/s
     |████████████████████████████████| 43 kB 1.2 MB/s 
     |████████████████████████████████| 376 kB 9.1 MB/s 
  Created wheel for kenlm: filename=kenlm-0.0.0-cp37-cp37m-linux_x86_64.whl size=2335801 sha256=34f5918b31ff0eb46feb619399879736f4aa81eb046692d6790971f8f9d8be9d
  Stored in directory: /tmp/pip-ephem-wheel-cache-vpasws9d/wheels/3d/aa/02/7b4a2eab5d7a2a9391bd9680dbad6270808a147bc3b7047e4e
  Created wheel for pygtrie: filename=pygtrie-2.4.2-py3-none-any.whl size=19063 sha256=f23afae85899200c0cb283b47f12b71ea8e3909ca50e8708ff66ae9c79d33d12
  Stored in directory: /root/.cache/pip/wheels/d3/f8/ba/1d828b1603ea422686eb694253a43cb3a5901ea4696c1e0603
Successfully built kenlm pygtrie


For demonstration purposes, we have prepared a new model repository [patrickvonplaten/wav2vec2-base-100h-with-lm](https://huggingface.co/patrickvonplaten/wav2vec2-base-100h-with-lm) which contains the same Wav2Vec2 checkpoint but has an additional **4-gram** language model for English.

Instead of using `Wav2Vec2Processor`, this time we use `Wav2Vec2ProcessorWithLM` to load the **4-gram** model in addition to the feature extractor and tokenizer.

In [ ]:
from transformers import Wav2Vec2ProcessorWithLM

# processor = Wav2Vec2ProcessorWithLM.from_pretrained("patrickvonplaten/wav2vec2-base-100h-with-lm")
processor = Wav2Vec2ProcessorWithLM.from_pretrained(repo_name)

ImportError: ignored

In constrast to decoding the audio without language model, the processor now directly receives the model's output `logits` instead of the `argmax(logits)` (called `predicted_ids`) above. The reason is that when decoding with a language model, at each time step, the processor takes the probabilities of all possible output characters into account. Let's take a look at the dimension of the `logits` output.

In [ ]:
logits.shape

torch.Size([1, 624, 32])

We can see that the `logits` correspond to a sequence of 624 vectors each having 32 entries. Each of the 32 entries thereby stands for the logit probability of one of the 32 possible output characters of the model:

In [ ]:
" ".join(sorted(processor.tokenizer.get_vocab()))

'</s> <s> [PAD] [UNK] | । ঁ ং ঃ অ আ ই ঈ উ এ ঐ ও ঔ ক খ গ ঘ ঙ চ ছ জ ঝ ঞ ট ঠ ড ঢ ণ ত থ দ ধ ন প ফ ব ভ ম য র ল শ ষ স হ ় া ি ী ু ূ ৃ ে ৈ ো ৌ ্ ৎ ড় ঢ় য় ৰ ৱ ’'

Intuitively, one can understand the decoding process of `Wav2Vec2ProcessorWithLM` as applying beam search through a matrix of size 624 $\times$ 32 probabilities while leveraging the probabilities of the next letters as given by the *n-gram* language model.

OK, let's run the decoding step again. `pyctcdecode` language model decoder does not automatically convert `torch` tensors to `numpy` so we'll have to convert them ourselves before.

In [ ]:
transcription = processor.batch_decode(logits.numpy()).text
transcription[0].lower()

'he tells us that at this festive season of the year with christmas and rose beef looming before us similes drawn from eating and its results occur most readily to the mind'

Cool! Recalling the words `facebook/wav2vec2-base-100h` without a language model transcribed incorrectly previously, *e.g.*,

> - *christmaus* vs. *christmas*
- *rose* vs. *roast*
- *simalyis* vs. *similes*

we can take another look at the transcription of `facebook/wav2vec2-base-100h` **with** a 4-gram language model. 2 out of 3 errors are corrected; *christmas* and *similes* have been correctly transcribed.

Interestingly, the incorrect transcription of *rose* persists. However, this should not surprise us very much. Decoding audio without a language model is much more prone to yield spelling mistakes, such as *christmaus* or *similes* (those words don't exist in the English language as far as I know). This is because the speech recognition system almost solely bases its prediction on the acoustic input it was given and not really on the language modeling context of previous and successive predicted letters ${}^1$. 
If on the other hand, we add a language model, we can be fairly sure that the speech recognition system will heavily reduce spelling errors since a well-trained *n-gram* model will surely not predict a word that has spelling errors. But the word *rose* is a valid English word and therefore the 4-gram will predict this word with a probability that is not insignificant. 

The language model on its own most likely does favor the correct word *roast* since the word sequence *roast beef* is much more common in English than *rose beef*. Because the final transcription is derived from a weighted combination of `facebook/wav2vec2-base-100h` output probabilities and those of the *n-gram* language model, it is quite common to see incorrectly transcribed words such as *rose*.

For more information on how you can tweak different parameters when decoding with `Wav2Vec2ProcessorWithLM`, please take a look at the official documentation [here](https://huggingface.co/docs/transformers/master/en/model_doc/wav2vec2#transformers.Wav2Vec2ProcessorWithLM.batch_decode).

---
${}^1$ Some research shows that a model such as `facebook/wav2vec2-base-100h` - when sufficiently large and trained on enough data - can learn language modeling dependencies between intermediate audio representations similar to a language model.


Great, now that you have seen the advantages adding an *n-gram* language model can bring, let's dive into how to create an *n-gram* and `Wav2Vec2ProcessorWithLM` from scratch.

## **2. Getting data for your language model**

A language model that is useful for a speech recognition system should support the acoustic model, *e.g.* Wav2Vec2, in predicting the next word (or token, letter) and therefore model the following distribution:

$\mathbf{P}(w_n | \mathbf{w}_0^{t-1})$ with $w_n$ being the next word and $\mathbf{w}_0^{t-1}$ being the sequence of all previous words since the beginning of the utterance. Simply said, the language model should be good at predicting the next word given all previously transcribed words regardless of the audio input given to the speech recognition system.

As always a language model is only as good as the data it is trained on. In the case of speech recognition, we should therefore ask ourselves for what kind of data, the speech recognition will be used for: *conversations*, *audiobooks*, *movies*, *speeches*, *, etc*, ...?

The language model should be good at modeling language that corresponds to the 
target transcriptions of the speech recognition system. 
For demonstration purposes, we assume here that we have fine-tuned a pre-trained [`facebook/wav2vec2-xls-r-300m`](https://huggingface.co/facebook/wav2vec2-xls-r-300m) on [Common Voice 7](https://huggingface.co/datasets/mozilla-foundation/common_voice_7_0) in Swedish. The fine-tuned checkpoint can 
be found [here](https://huggingface.co/hf-test/xls-r-300m-sv).
Common Voice 7 is a relatively crowd-sourced read-out audio dataset and we will evaluate the model on its test data.

Let's now look for suitable text data on the Hugging Face Hub. We search all datasets for those [that contain Swedish data](https://huggingface.co/datasets?languages=languages:sv&sort=downloads). 
Browsing a bit through the datasets, we are looking for a dataset that is similar to Common Voice's read-out audio data. The obvious choices of [oscar](https://huggingface.co/datasets/oscar) and [mc4](https://huggingface.co/datasets/mc4) might not be the most suitable here because they:

- are generated from crawling the web, which might not be very clean and correspond well to spoken language
- require a lot of pre-processing
- are very large which is not ideal for demonstration purposes here 😉

A dataset that seems sensible here and which is relatively clean and easy to pre-process is [europarl_bilingual](https://huggingface.co/datasets/europarl_bilingual) as it's a dataset that is based on discussions and talks of the European parliament. It should therefore be relatively clean and correspond well to read-out audio data. The dataset is originally designed for machine translation and can therefore only be accessed in translation pairs. We will only extract the text of the target language, Swedish (`sv`), from the *English-to-Swedish* translations.

In [ ]:
target_lang="as"  # change to your target lang

Let's download the data.

In [ ]:
from datasets import load_dataset

dataset = load_dataset("europarl_bilingual", lang1="en", lang2=target_lang, split="train")
dataset

Downloading:   0%|          | 0.00/2.60k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/29.6k [00:00<?, ?B/s]

Using custom data configuration en-sv-lang1=en,lang2=sv


Downloading:   0%|          | 0.00/142M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/127M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/8.90M [00:00<?, ?B/s]

0 examples [00:00, ? examples/s]

Dataset europarl_bilingual downloaded and prepared to /root/.cache/huggingface/datasets/europarl_bilingual/en-sv-lang1=en,lang2=sv/8.0.0/2ab0200e7729616bfd4a4df6bfb29b31746ceb5a59f8c75c02ca35e1ebead950. Subsequent calls will reuse this data.


We see that the data is quite large - it has over a million translations. Since it's only text data, it should be relatively easy to process though.

Next, let's look at how the data was preprocessed when training the fine-tuned *XLS-R* checkpoint in Swedish. Looking at the [`run.sh` file](https://huggingface.co/hf-test/xls-r-300m-sv/blob/main/run.sh), we can see that the following characters were removed from the official transcriptions:

In [ ]:
chars_to_remove_regex = '[\,\?\.\!\-\;\:\"\“\%\‘\”\�\']'  # change to the ignored characters of your fine-tuned model

Let's do the same here so that the alphabet of our language model matches the one of the fine-tuned acoustic checkpoints.

We can write a single map function to extract the Swedish text and process it right away.

In [ ]:
import re

def extract_text(batch):
  text = batch["translation"][target_lang]
  batch["text"] = re.sub(chars_to_ignore_regex, "", text.lower())
  return batch

Let's apply the `.map()` function. This should take roughly 5 minutes.

In [ ]:
dataset = dataset.map(extract_text, remove_columns=dataset.column_names)
dataset

  0%|          | 0/281 [00:00<?, ?ex/s]

KeyError: ignored

Great. Our dataset is already finished. Let's upload it to the Hub so that we can inspect and reuse it better.

You can log in by executing the following cell.

In [ ]:
from huggingface_hub import notebook_login

notebook_login()

Login successful
Your token has been saved to /root/.huggingface/token
Authenticated through git-credential store but this isn't the helper defined on your machine.
You might have to re-authenticate when pushing to the Hugging Face Hub. Run the following command in your terminal in case you want to set this credential helper as the default

git config --global credential.helper store


Next, we call 🤗 Hugging Face's [`push_to_hub`](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=push#datasets.Dataset.push_to_hub) method to upload the dataset to the repo `"swedish_corpora_parliament_processed"`.

In [ ]:
dataset.push_to_hub(f"{target_lang}_corpora_parliament_processed", split="train")

The repository already exists: the `private` keyword argument will be ignored.


Pushing dataset shards to the dataset hub:   0%|          | 0/1 [00:00<?, ?it/s]

That was easy! The dataset viewer is automatically enabled when uploading a new dataset, which is very convenient. You can now directly inspect the dataset online. 

Feel free to look through our preprocessed dataset directly on [`hf-test/sv_corpora_parliament_processed`](https://huggingface.co/datasets/hf-test/sv_corpora_parliament_processed). Even if we are not a native speaker in Swedish, we can see that the data is well processed and seems clean.

Next, let's use the data to build a language model.

## **3. Build an *n-gram* with KenLM**

While large language models based on the [Transformer architecture](https://jalammar.github.io/illustrated-transformer/) have become the standard in NLP, it is still very common to use an ***n-gram*** LM to boost speech recognition systems - as shown in Section 1.

Looking again at Table 9 of Appendix C of the [official Wav2Vec2 paper](https://arxiv.org/abs/2006.11477), it can be noticed that using a *Transformer*-based LM for decoding clearly yields better results than using an *n-gram* model, but the difference between *n-gram* and *Transformer*-based LM is much less significant than the difference between *n-gram* and no LM. 

*E.g.*, for the large Wav2Vec2 checkpoint that was fine-tuned on 10min only, an *n-gram* reduces the word error rate (WER) compared to no LM by *ca.* 80% while a *Transformer*-based LM *only* reduces the WER by another 23% compared to the *n-gram*. This relative WER reduction becomes less, the more data the acoustic model has been trained on. *E.g.*, for the large checkpoint a *Transformer*-based LM reduces the WER by merely 8% compared to an *n-gram* LM whereas the *n-gram* still yields a 21% WER reduction compared to no language model.

The reason why an *n-gram* is preferred over a *Transformer*-based LM is that *n-grams* come at a significantly smaller computational cost. For an *n-gram*, retrieving the probability of a word given previous words is almost only as computationally expensive as querying a look-up table or tree-like data storage - *i.e.* it's very fast compared to modern *Transformer*-based language models that would require a full forward pass to retrieve the next word probabilities.

For more information on how *n-grams* function and why they are (still) so useful for speech recognition, the reader is advised to take a look at [this excellent summary](https://web.stanford.edu/~jurafsky/slp3/3.pdf) from Stanford.

Great, let's see step-by-step how to build an *n-gram*. We will use the popular [KenLM library](https://github.com/kpu/kenlm) to do so. Let's start by installing the Ubuntu library prerequisites:

In [ ]:
!sudo apt install build-essential cmake libboost-system-dev libboost-thread-dev libboost-program-options-dev libboost-test-dev libeigen3-dev zlib1g-dev libbz2-dev liblzma-dev

Reading package lists... Done
Building dependency tree       
Reading state information... Done
build-essential is already the newest version (12.4ubuntu1).
libboost-program-options-dev is already the newest version (1.65.1.0ubuntu1).
libboost-program-options-dev set to manually installed.
libboost-system-dev is already the newest version (1.65.1.0ubuntu1).
libboost-system-dev set to manually installed.
libboost-thread-dev is already the newest version (1.65.1.0ubuntu1).
libboost-thread-dev set to manually installed.
liblzma-dev is already the newest version (5.2.2-1.3).
liblzma-dev set to manually installed.
zlib1g-dev is already the newest version (1:1.2.11.dfsg-0ubuntu2).
zlib1g-dev set to manually installed.
libboost-test-dev is already the newest version (1.65.1.0ubuntu1).
libboost-test-dev set to manually installed.
cmake is already the newest version (3.10.2-1ubuntu2.18.04.2).
libbz2-dev is already the newest version (1.0.6-8.1ubuntu0.2).
libbz2-dev set to manually installed.
Su

before downloading and unpacking the KenLM repo.

In [ ]:
!wget -O - https://kheafield.com/code/kenlm.tar.gz | tar xz

--2022-01-21 15:13:42--  https://kheafield.com/code/kenlm.tar.gz
Resolving kheafield.com (kheafield.com)... 35.196.63.85
Connecting to kheafield.com (kheafield.com)|35.196.63.85|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 491090 (480K) [application/x-gzip]
Saving to: ‘STDOUT’

-                   100%[===================>] 479.58K  --.-KB/s    in 0.02s   

2022-01-21 15:13:42 (18.9 MB/s) - written to stdout [491090/491090]



KenLM is written in C++, so we'll make use of `cmake` to build the binaries.

In [ ]:
!mkdir kenlm/build && cd kenlm/build && cmake .. && make -j2
!ls kenlm/build/bin

-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.65.1
-- Found the following Boost libraries:


Great, as we can see, the executable functions have successfully been built under `kenlm/build/bin/`.

KenLM by default computes an *n-gram* with [Kneser-Ney smooting](https://en.wikipedia.org/wiki/Kneser%E2%80%93Ney_smoothing). All text data used to create the *n-gram* is expected to be stored in a text file.
We download our dataset and save it as a `.txt` file.

In [ ]:
# from datasets import load_dataset

# username = "hf-test"  # change to your username

# dataset = load_dataset(f"{username}/{target_lang}_corpora_parliament_processed", split="train")

# with open("text.txt", "w") as file:
#   file.write(" ".join(dataset["text"]))

reference https://indicnlp.ai4bharat.org/corpora/

In [ ]:
!wget https://storage.googleapis.com/ai4bharat-public-indic-nlp-corpora/indiccorp/as.tar.xz

--2022-01-21 09:55:59--  https://storage.googleapis.com/ai4bharat-public-indic-nlp-corpora/indiccorp/as.tar.xz
Resolving storage.googleapis.com (storage.googleapis.com)... 142.251.16.128, 172.217.0.48, 172.217.1.208, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|142.251.16.128|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 66709392 (64M) [application/x-tar]
Saving to: ‘as.tar.xz’

as.tar.xz           100%[===================>]  63.62M  35.3MB/s    in 1.8s    

2022-01-21 09:56:02 (35.3 MB/s) - ‘as.tar.xz’ saved [66709392/66709392]



In [ ]:
!tar -xf as.tar.xz

In [ ]:
!head ./data/as/as.txt

NE INDIA: হোজাই জিলাৰ লংকাৰ কাকীত বুধবাৰে নিশা স্থানীয় ৰাইজে এটা উট সহ দুজন ব্যক্তিক আটক কৰি গতাই দিয়ে কাকী আৰক্ষীক | আৰক্ষীয়ে আটকাধীন ব্যক্তি দুজনৰ ওপৰত তদন্ত অব্যাহত ৰাখিছে।
ইয়াৰে প্ৰায় ১৫জন যাত্ৰী আহত হৈছে।
প্রতিটো স্তৰত ৬টা কৈ সৰু ফুল ফুলে ।
উল্লেখ্য যে অহাবছৰ এই কেন্দ্ৰটোৰ উদ্যোগতে গুৱাহাটীত ৰাষ্ট্ৰীয় জুনিয়ৰ ইকুৱেষ্ট্ৰিয়ান প্ৰতিযোগিতৰো আয়োজনৰ পৰিকল্পনা হাতত লোৱা হৈছে।
Home > কহঁৰা মধ্য ইংৰাজী বিদ্যালয়ৰ ছাত্ৰৰ মৃত্যু
সাংবাদিকে বুঢ়া পাগল হোৱা বুলি আঁতৰি গ’ল।
কাৰণ যিকোনো এটা জনগোষ্ঠীয়ে তেওঁলোকৰ জনসংখ্যাৰ ভিত্তিত ভাষা ব্যৱহাৰৰ দাবী কৰিব পাৰে৷ এই দিশত আমি কেৱল ব্রহ্মপুত্র উপত্যকাৰ কথা ভাবিলেই নহ’ব৷ ভাবিব লাগিব বৰাক উপত্যকাৰ কথাও৷ সংখ্যাগত দিশৰ কথা আহিলেই সেই লোকসকলে বৰাক উপত্যকাকো সামৰি ল’ব৷ যদিহে তেনেকুৱা পৰিস্থিতিৰ সৃষ্টি হয় তেন্তে অসমত বাংলাভাষীৰ সংখ্যা কিমান হ’বগৈ?
‘কিন্তু তাৰ পিছতো এয়া কোনেও লক্ষ্য কৰা নাই!
দেশত ভাল শিক্ষকৰ সাংঘাতিক অভাৱ হৈছে৷ সাম্প্রতিক সময়ত নতুন শিক্ষানীতিৰ সমান্তৰালভাৱে ভাল শিক্ষক গঢ়াটোও গুৰুত্বপূর্ণ হৈ পৰিছে৷– এই মন্তব্য ডিব্রুগড় বিশ্ববিদ্যালয়ৰ প্রাক্তন উপাচ

In [ ]:
%%time
!sed 's/[a-zA-Z0-9>:]//g' ./data/as/as.txt > as_cleaned.txt
!sed 's/।/ । /g' ./as_cleaned.txt > as_cleaned.1.txt
!mv as_cleaned.1.txt text.txt
!ls -ltr

total 1080940
-rw-r--r-- 1 root root  66709392 Sep 23  2020 as.tar.xz
drwxr-xr-x 1 root root      4096 Jan  7 14:33 sample_data
drwxr-xr-x 9 1000 1000      4096 Jan 21 09:54 kenlm
drwxr-xr-x 3 root root      4096 Jan 21 09:56 data
-rw-r--r-- 1 root root 519100593 Jan 21 10:19 as_cleaned.txt
-rw-r--r-- 1 root root 521052779 Jan 21 10:19 text.txt
CPU times: user 3.15 s, sys: 560 ms, total: 3.71 s
Wall time: 7min 7s


Now, we just have to run KenLM's `lmplz` command to build our *n-gram*, called `"5gram.arpa"`. As it's relatively common in speech recognition, we build a *5-gram* by passing the `-o 5` parameter.
For more information on the different *n-gram* LM that can be built 
with KenLM, one can take a look at the [official website of KenLM](https://kheafield.com/code/kenlm/).

Executing the command below might take a minute or so.

In [ ]:

!kenlm/build/bin/lmplz -o 5 <"text.txt" > "5gram.arpa"

=== 1/5 Counting and sorting n-grams ===
Reading /content/text.txt
----5---10---15---20---25---30---35---40---45---50---55---60---65---70---75---80---85---90---95--100
tcmalloc: large alloc 1918697472 bytes == 0x55c12394c000 @  0x7f995294e1e7 0x55c1220157a2 0x55c121fb051e 0x55c121f8f2eb 0x55c121f7b066 0x7f9950ae7bf7 0x55c121f7cbaa
tcmalloc: large alloc 8953896960 bytes == 0x55c195f1c000 @  0x7f995294e1e7 0x55c1220157a2 0x55c1220047ca 0x55c122005208 0x55c121f8f308 0x55c121f7b066 0x7f9950ae7bf7 0x55c121f7cbaa
****************************************************************************************************
Unigram tokens 30408490 types 1100270
=== 2/5 Calculating and sorting adjusted counts ===
Chain sizes: 1:13203240 2:1061907136 3:1991075968 4:3185721344 5:4645843968
tcmalloc: large alloc 4645847040 bytes == 0x55c12394c000 @  0x7f995294e1e7 0x55c1220157a2 0x55c1220047ca 0x55c122005208 0x55c121f8f8d7 0x55c121f7b066 0x7f9950ae7bf7 0x55c121f7cbaa
tcmalloc: large alloc 1991081984 bytes =

Great, we have built a *5-gram* LM! Let's inspect the first couple of lines.

In [ ]:
!head -20 5gram.arpa

\data\
ngram 1=1100270
ngram 2=10196386
ngram 3=20202858
ngram 4=24388635
ngram 5=25279719

\1-grams:
-7.0538974	<unk>	0
0	<s>	-1.0157648
-2.1170783	</s>	0
-4.0849	হোজাই	-0.7007276
-3.6245263	জিলাৰ	-0.3177578
-4.660954	লংকাৰ	-0.24585101
-6.026935	কাকীত	-0.11453828
-3.6779141	বুধবাৰে	-0.4127841
-3.7027547	নিশা	-0.3863523
-3.4385362	স্থানীয়	-0.5747764
-3.4661708	ৰাইজে	-0.4619786
-2.8084285	এটা	-0.53903985


There is a small problem that 🤗 Transformers will not be happy about later on.
The *5-gram* correctly includes a "Unknown" or `<unk>`, as well as a *begin-of-sentence*, `<s>` token, but no *end-of-sentence*, `</s>` token.
This sadly has to be corrected currently after the build.

We can simply add the *end-of-sentence* token by adding the line `0 </s>  -0.11831701` below the *begin-of-sentence* token and increasing the `ngram 1` count by 1. Because the file has roughly 100 million lines, this command will take *ca.* 2 minutes.

In [ ]:
with open("5gram.arpa", "r") as read_file, open("5gram_correct.arpa", "w") as write_file:
  has_added_eos = False
  for line in read_file:
    if not has_added_eos and "ngram 1=" in line:
      count=line.strip().split("=")[-1]
      write_file.write(line.replace(f"{count}", f"{int(count)+1}"))
    elif not has_added_eos and "<s>" in line:
      write_file.write(line)
      write_file.write(line.replace("<s>", "</s>"))
      has_added_eos = True
    else:
      write_file.write(line)

Let's now inspect the corrected *5-gram*.

In [ ]:
!head -20 5gram_correct.arpa

\data\
ngram 1=1100271
ngram 2=10196386
ngram 3=20202858
ngram 4=24388635
ngram 5=25279719

\1-grams:
-7.0538974	<unk>	0
0	<s>	-1.0157648
0	</s>	-1.0157648
-2.1170783	</s>	0
-4.0849	হোজাই	-0.7007276
-3.6245263	জিলাৰ	-0.3177578
-4.660954	লংকাৰ	-0.24585101
-6.026935	কাকীত	-0.11453828
-3.6779141	বুধবাৰে	-0.4127841
-3.7027547	নিশা	-0.3863523
-3.4385362	স্থানীয়	-0.5747764
-3.4661708	ৰাইজে	-0.4619786


Great, this looks better! We're done at this point and all that is left to do is to correctly integrate the `"ngram"` with [`pyctcdecode`](https://github.com/kensho-technologies/pyctcdecode) and 🤗 Transformers.

## **4. Combine an *n-gram* with Wav2Vec2**

In a final step, we want to wrap the *5-gram* into a `Wav2Vec2ProcessorWithLM` object to make the *5-gram* boosted decoding as seamless as shown in Section 1.
We start by downloading the currently "LM-less" processor of [`xls-r-300m-sv`](https://huggingface.co/hf-test/xls-r-300m-sv).

In [ ]:
from transformers import AutoProcessor

processor = AutoProcessor.from_pretrained(repo_name)

Next, we extract the vocabulary of its tokenizer as it represents the `"labels"` of `pyctcdecode`'s `BeamSearchDecoder` class.

In [ ]:
vocab_dict = processor.tokenizer.get_vocab()
sorted_vocab_dict = {k.lower(): v for k, v in sorted(vocab_dict.items(), key=lambda item: item[1])}

The `"labels"` and the previously built `5gram_correct.arpa` file is all that's needed to build the decoder. 

In [ ]:
ls -lltr

total 12
drwxr-xr-x 1 root root 4096 Jan  7 14:33 sample_data/
drwxr-xr-x 9 1000 1000 4096 Jan 21 15:13 kenlm/
-rw-r--r-- 1 root root  135 Jan 21 15:16 5gram_correct.arpa


In [ ]:
!wget https://huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese/resolve/main/language_model/5gram_correct.arpa

--2022-01-21 15:18:50--  https://huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese/resolve/main/language_model/5gram_correct.arpa
Resolving huggingface.co (huggingface.co)... 34.204.221.201, 52.22.206.115, 34.224.55.150, ...
Connecting to huggingface.co (huggingface.co)|34.204.221.201|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cdn-lfs.huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese/ea017e6c759749e37cb72d9d2ba1f4a55502b0f19a903c6ab4ffe8429eedff89 [following]
--2022-01-21 15:18:50--  https://cdn-lfs.huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese/ea017e6c759749e37cb72d9d2ba1f4a55502b0f19a903c6ab4ffe8429eedff89
Resolving cdn-lfs.huggingface.co (cdn-lfs.huggingface.co)... 13.226.52.14, 13.226.52.13, 13.226.52.128, ...
Connecting to cdn-lfs.huggingface.co (cdn-lfs.huggingface.co)|13.226.52.14|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6978078575 (6.5G) [binary/octet-stream]
Sav

In [ ]:
!mv 5gram_correct.arpa.1 5gram_correct.arpa

In [ ]:
!head 5gram_correct.arpa

\data\
ngram 1=1100271
ngram 2=10196386
ngram 3=20202858
ngram 4=24388635
ngram 5=25279719

\1-grams:
-7.0538974	<unk>	0
0	<s>	-1.0157648


In [ ]:
from pyctcdecode import build_ctcdecoder

decoder = build_ctcdecoder(
    labels=list(sorted_vocab_dict.keys()),
    kenlm_model_path="5gram_correct.arpa",
)

Found entries of length > 1 in alphabet. This is unusual unless style is BPE, but the alphabet was not recognized as BPE type. Is this correct?
Unigrams and labels don't seem to agree.


We can safely ignore the warning and all that is left to do now is to wrap the just created `decoder`, together with the processor's `tokenizer` and `feature_extractor` into a `Wav2Vec2ProcessorWithLM` class.

In [ ]:
from transformers import Wav2Vec2ProcessorWithLM

processor_with_lm = Wav2Vec2ProcessorWithLM(
    feature_extractor=processor.feature_extractor,
    tokenizer=processor.tokenizer,
    decoder=decoder
)

We want to directly upload the LM-boosted processor into 
the model folder of [`xls-r-300m-sv`](https://huggingface.co/hf-test/xls-r-300m-sv) to have all relevant files in one place.

Let's clone the repo, add the new decoder files and upload them afterward.
First, we need to install `git-lfs`.

In [ ]:
!sudo apt-get install git-lfs tree

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  git-lfs tree
0 upgraded, 2 newly installed, 0 to remove and 37 not upgraded.
Need to get 2,169 kB of archives.
After this operation, 7,767 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 git-lfs amd64 2.3.4-1 [2,129 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 tree amd64 1.7.0-5 [40.7 kB]
Fetched 2,169 kB in 1s (2,415 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 2.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Select

Cloning and uploading of modeling files can be done conveniently with the `huggingface_hub`'s `Repository` class. 

More information on how to use the `huggingface_hub` to upload any files, please take a look at the [official docs](https://huggingface.co/docs/hub/how-to-upstream).

In [ ]:
print(repo_name)

infinitejoy/wav2vec2-large-xls-r-300m-assamese


In [ ]:
from huggingface_hub import Repository

repo = Repository(local_dir=repo_name, clone_from=repo_name)

Cloning https://huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese into local empty directory.


Download file pytorch_model.bin:   0%|          | 16.0k/1.18G [00:00<?, ?B/s]

Download file runs/Jan19_14-02-34_Yellow/events.out.tfevents.1642581173.Yellow.1965976.0:  68%|######7   | 3.4…

Download file runs/Jan19_14-02-34_Yellow/1642581173.3815742/events.out.tfevents.1642581173.Yellow.1965976.1:  …

Download file runs/Jan19_14-33-09_Yellow/1642583005.5649233/events.out.tfevents.1642583005.Yellow.1965976.3:  …

Download file runs/Jan18_11-45-41_Yellow/events.out.tfevents.1642486556.Yellow.1854158.6:  48%|####8     | 3.4…

Download file runs/Jan18_11-45-41_Yellow/1642486556.2523344/events.out.tfevents.1642486556.Yellow.1854158.7:  …

Download file runs/Jan19_14-33-09_Yellow/events.out.tfevents.1642583005.Yellow.1965976.2: 100%|##########| 10.…

Download file runs/Jan18_10-33-31_Yellow/events.out.tfevents.1642482228.Yellow.1854158.2:  59%|#####9    | 3.4…

Clean file runs/Jan19_14-02-34_Yellow/events.out.tfevents.1642581173.Yellow.1965976.0:  20%|#9        | 1.00k/…

Download file runs/Jan18_11-07-35_Yellow/1642484270.371665/events.out.tfevents.1642484270.Yellow.1854158.5:  7…

Clean file runs/Jan19_14-33-09_Yellow/1642583005.5649233/events.out.tfevents.1642583005.Yellow.1965976.3:  21%…

Clean file runs/Jan19_14-02-34_Yellow/1642581173.3815742/events.out.tfevents.1642581173.Yellow.1965976.1:  21%…

Download file runs/Jan18_10-29-03_Yellow/events.out.tfevents.1642481962.Yellow.1854158.0:  85%|########4 | 3.4…

Clean file runs/Jan19_14-33-09_Yellow/events.out.tfevents.1642583005.Yellow.1965976.2:   9%|9         | 1.00k/…

Clean file runs/Jan18_11-45-41_Yellow/1642486556.2523344/events.out.tfevents.1642486556.Yellow.1854158.7:  22%…

Download file training_args.bin: 100%|##########| 2.98k/2.98k [00:00<?, ?B/s]

Download file runs/Jan18_10-33-31_Yellow/1642482228.516824/events.out.tfevents.1642482228.Yellow.1854158.3: 10…

Clean file runs/Jan18_11-45-41_Yellow/events.out.tfevents.1642486556.Yellow.1854158.6:  14%|#3        | 1.00k/…

Download file runs/Jan18_11-07-35_Yellow/events.out.tfevents.1642484270.Yellow.1854158.4:  85%|########4 | 3.4…

Clean file runs/Jan18_10-33-31_Yellow/events.out.tfevents.1642482228.Yellow.1854158.2:  17%|#7        | 1.00k/…

Download file runs/Jan18_10-29-03_Yellow/1642481962.2863166/events.out.tfevents.1642481962.Yellow.1854158.1:  …

Clean file runs/Jan18_11-07-35_Yellow/1642484270.371665/events.out.tfevents.1642484270.Yellow.1854158.5:  22%|…

Clean file runs/Jan18_10-29-03_Yellow/events.out.tfevents.1642481962.Yellow.1854158.0:  24%|##4       | 1.00k/…

Clean file training_args.bin:  34%|###3      | 1.00k/2.98k [00:00<?, ?B/s]

Clean file runs/Jan18_10-33-31_Yellow/1642482228.516824/events.out.tfevents.1642482228.Yellow.1854158.3:  22%|…

Clean file runs/Jan18_11-07-35_Yellow/events.out.tfevents.1642484270.Yellow.1854158.4:  24%|##4       | 1.00k/…

Clean file runs/Jan18_10-29-03_Yellow/1642481962.2863166/events.out.tfevents.1642481962.Yellow.1854158.1:  22%…

Clean file pytorch_model.bin:   0%|          | 1.00k/1.18G [00:00<?, ?B/s]

Having cloned `xls-r-300m-sv`, let's save the new processor with LM into it.

In [ ]:
processor_with_lm.save_pretrained(repo_name)

Let's inspect the local repository. The `tree` command conveniently can also show the size of the different files.

In [ ]:
!tree -h infinitejoy/wav2vec2-large-xls-r-300m-assamese

infinitejoy/wav2vec2-large-xls-r-300m-assamese
├── [  23]  added_tokens.json
├── [ 703]  alphabet.json
├── [2.0K]  config.json
├── [4.0K]  language_model
│   ├── [6.5G]  5gram_correct.arpa
│   ├── [  78]  attrs.json
│   └── [ 27M]  unigrams.txt
├── [ 214]  preprocessor_config.json
├── [1.2G]  pytorch_model.bin
├── [2.2K]  README.md
├── [4.0K]  runs
│   ├── [4.0K]  Jan18_10-29-03_Yellow
│   │   ├── [4.0K]  1642481962.2863166
│   │   │   └── [4.5K]  events.out.tfevents.1642481962.Yellow.1854158.1
│   │   └── [4.1K]  events.out.tfevents.1642481962.Yellow.1854158.0
│   ├── [4.0K]  Jan18_10-33-31_Yellow
│   │   ├── [4.0K]  1642482228.516824
│   │   │   └── [4.5K]  events.out.tfevents.1642482228.Yellow.1854158.3
│   │   └── [5.9K]  events.out.tfevents.1642482228.Yellow.1854158.2
│   ├── [4.0K]  Jan18_11-07-35_Yellow
│   │   ├── [4.0K]  1642484270.371665
│   │   │   └── [4.5K]  events.out.tfevents.1642484270.Yellow.1854158.5
│   │   └── [4.1K]  events.out.tfevents.1642484270.Yellow.1854158.4


As can be seen the *5-gram* LM is quite large - it amounts to more than 4 GB.
To reduce the size of the *n-gram* and make loading faster, `kenLM` allows converting `.arpa` files to binary ones using the `build_binary` executable.

Let's make use of it here.

In [ ]:
!kenlm/build/bin/build_binary infinitejoy/wav2vec2-large-xls-r-300m-assamese/language_model/5gram_correct.arpa xls-r-300m-sv/language_model/5gram.bin

Reading infinitejoy/wav2vec2-large-xls-r-300m-assamese/language_model/5gram_correct.arpa
----5---10---15---20---25---30---35---40---45---50---55---60---65---70---75---80---85---90---95--100
****************************************************************************************************
/content/kenlm/util/file.cc:86 in int util::CreateOrThrow(const char*) threw ErrnoException because `-1 == (ret = open(name, 0100 | 01000 | 02, 0400 | 0200 | (0400 >> 3) | ((0400 >> 3) >> 3)))'.
No such file or directory while creating xls-r-300m-sv/language_model/5gram.bin Byte: 92
ERROR


Great, it worked! Let's remove the `.arpa` file and check the size of the binary *5-gram* LM.

In [ ]:
!rm xls-r-300m-sv/language_model/5gram_correct.arpa && tree -h xls-r-300m-sv/

xls-r-300m-sv/
├── [  23]  added_tokens.json
├── [ 401]  all_results.json
├── [ 253]  alphabet.json
├── [2.0K]  config.json
├── [ 304]  emissions.csv
├── [ 226]  eval_results.json
├── [4.0K]  language_model
│   ├── [1.8G]  5gram.bin
│   ├── [  78]  attrs.json
│   └── [4.9M]  unigrams.txt
├── [ 240]  preprocessor_config.json
├── [1.2G]  pytorch_model.bin
├── [3.5K]  README.md
├── [4.0K]  runs
│   └── [4.0K]  Jan09_22-00-50_brutasse
│       ├── [4.0K]  1641765760.8871996
│       │   └── [4.6K]  events.out.tfevents.1641765760.brutasse.31164.1
│       ├── [ 42K]  events.out.tfevents.1641765760.brutasse.31164.0
│       └── [ 364]  events.out.tfevents.1641794162.brutasse.31164.2
├── [1.2K]  run.sh
├── [ 30K]  run_speech_recognition_ctc.py
├── [ 502]  special_tokens_map.json
├── [ 279]  tokenizer_config.json
├── [ 29K]  trainer_state.json
├── [2.9K]  training_args.bin
├── [ 196]  train_results.json
├── [ 319]  vocab.json
└── [4.0K]  wandb
    ├── [  52]  debug-internal.log -> run-20220109_220

Nice, we reduced the *n-gram* by more than half to less than 2GB now. In the final step, let's upload all files.

In [ ]:
repo.push_to_hub(commit_message="Upload lm-boosted decoder")

Adding files tracked by Git LFS: ['language_model/5gram_correct.arpa', 'language_model/unigrams.txt']. This may take a bit of time if the files are large.


Upload file language_model/5gram_correct.arpa:   0%|          | 1.00/6.50G [00:00<?, ?B/s]

Upload file language_model/unigrams.txt:   0%|          | 1.00/27.4M [00:00<?, ?B/s]

To https://huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese
   90d66ba..1113356  main -> main



'https://huggingface.co/infinitejoy/wav2vec2-large-xls-r-300m-assamese/commit/11133563cc4b1c2cc0dab18fd282c12cf14cb9ce'

That's it. Now you should be able to use the *5gram* for LM-boosted decoding as shown in Section 1.

As can be seen on [`xls-r-300m-sv`'s model card](https://huggingface.co/hf-test/xls-r-300m-sv#inference-with-lm) our *5gram* LM-boosted decoder yields a WER of 18.85% on Common Voice's 7 test set which is a relative performance of *ca.* 30% 🔥.